config/confignet/confignet.go

```diff
@@ -36,20 +84,32 @@ type NetAddr struct {
 // Dial equivalent with net.Dialer's DialContext for this address.
 func (na *NetAddr) Dial(ctx context.Context) (net.Conn, error) {
 	d := net.Dialer{Timeout: na.DialerConfig.Timeout}
-	return d.DialContext(ctx, na.Transport, na.Endpoint)
+
+	tt := string(na.TransportType)
+	if na.Transport != "" {
+		tt = na.Transport
+	}
```

Comment:
I think there is a bug here. The reason is that implementations may set the default `Transport` to `"foo"` then user sets it to `"bar"`, based on this logic if I am not mistaken you will end up with `"foo"`.

So in order to make it work (I think we can still do it if you want to not just break the go API) with implementing the `Unmarshaler` interface and before unmarshalling set the `TransportType` to the value of `Transport` then here you can ignore old value. Something like:

```golang
func (na *NetAddr) Unmarshal(confmap.Confmap cm) error {
  na.TrasportType = TransportType(na.Trasport)
  return cm.Unmarshal(na)
}
```

Reply:
What I tried didn't work for cases like these: https://github.com/open-telemetry/opentelemetry-collector/blob/1a57837accdff8bf9abc769a8e49b5207b52fad2/config/configgrpc/configgrpc_test.go#L618-L624 where a user is setting the public members directly, bypassing any unmashaling.

I've reverted the code to do a check for `na.Transport` in `Dial` and `Listen`. Can you explain more how that check introduces a bug for those using `Transport` instead of `TransportType`?

Reply:
@bogdandrutu can you explain further?

Reply:
User in their "defaultConfig"

```
return { Transport: "unix" } // configure the deprecated value (until I change the config)
```

In the yaml file:

```
  "transport": "tcp",
``` 

Then now will Call `Dial`, which protocol will be used?

Reply:
Oh I see what you mean.  I agree that is a bug.

Your proposed fix allows the default config being overridden by user config, but since `Transport` and `TransportType` are both public, users could still explicitly set `Transport` somewhere other than in via the default config function, meaning we'd still have to check for `Transport` being set in `Dial` and `Listen`.

Since this would only be a breaking change in the API and not for end user configs, I am now very much in favor of doing this change as a breaking API change.  I've update the PR to reflect that opinion.